lib/PersonalityProvider.jsm

```diff
@@ -0,0 +1,198 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+const {PersistentCache} = ChromeUtils.import("resource://activity-stream/lib/PersistentCache.jsm", {});
+const {RemoteSettings} = ChromeUtils.import("resource://services-settings/remote-settings.js", {});
+
+const {NaiveBayesTextTagger} = ChromeUtils.import("resource://activity-stream/lib/NaiveBayesTextTagger.jsm", {});
+const {NmfTextTagger} = ChromeUtils.import("resource://activity-stream/lib/NmfTextTagger.jsm", {});
+const {RecipeExecutor} = ChromeUtils.import("resource://activity-stream/lib/RecipeExecutor.jsm", {});
+
+ChromeUtils.defineModuleGetter(this, "NewTabUtils",
+  "resource://gre/modules/NewTabUtils.jsm");
+
+const STORE_UPDATE_TIME = 24 * 60 * 60 * 1000; // 24 hours
+
+/**
+ * V2 provider builds and ranks an interest profile (also called an “interest vector”) off the browse history.
+ * This allows Firefox to classify pages into topics, by examining the text found on the page.
+ * It does this by looking at the history text content, title, and description.
+ */
+this.PersonalityProvider = class PersonalityProvider {
+  constructor(
+    timeSegments,
+    parameterSets,
+    maxHistoryQueryResults,
+    version,
+    scores,
+    modelKeys) {
+    this.modelKeys = modelKeys;
+    this.timeSegments = timeSegments;
+    this.parameterSets = parameterSets;
+    this.maxHistoryQueryResults = maxHistoryQueryResults;
+    this.version = version;
+    this.scores = scores;
+    this.store = new PersistentCache("personality-provider", true);
+  }
+
+  async init() {
+    this.interestConfig = await this.getRecipe();
+    if (!this.interestConfig) {
+      return;
+    }
+    this.recipeExecutor = await this.generateRecipeExecutor();
+    if (!this.recipeExecutor) {
+      return;
+    }
+    this.interestVector = await this.store.get("interest-vector");
+
+    // Fetch a new one if none exists or every set update time.
+    if (!this.interestVector ||
+      (Date.now() - this.interestVector.lastUpdate) >= STORE_UPDATE_TIME) {
+      this.interestVector = await this.createInterestVector();
+      if (!this.interestVector) {
+        return;
+      }
+      this.interestVector.lastUpdate = Date.now();
+      this.store.set("interest-vector", this.interestVector);
+    }
+    this.initialized = true;
+  }
+
+  async getFromRemoteSettings(name) {
+    const result = await RemoteSettings(name).get();
+    return result || [];
+  }
+
+  /**
+   * Returns a Recipe from remote settings to be consumed by a RecipeExecutor.
+   * A Recipe is a set of instructions on how to processes a RecipeExecutor.
+   */
+  async getRecipe() {
+    if (!this.recipe) {
```

Comment:
Hmm, looks like there is a corner case here, assume that the first call of `this.recipe = await this.getFromRemoteSettings("personality-provider-recipe");` fails somehow, hence `this.recipe` will be an empty array. Then that user switches to the old version of personalization temporarily, then switches it back to V2 again. That will call the `init` function, which in turn calls `getRecipe()`. At this point, `this.recipe` is still an empty array, so it will skip the if statement without retrying the `RemoteSettings.get()`, which might be ready to query now.

Although this is rare, perhaps change it to `if (this.recipe === undefined || !this.recipe.length)`?

Reply:
How about `if (!this.recipe || !this.recipe.length)`?

Or was the `=== undefined` there to catch something specific?

Reply:
but this is a good catch. The if statement there was actually an early quick addition. Should probably be a bit more sophisticated, I just lost it in the shuffle.